### PR TITLE
feat: cors filter

### PR DESCRIPTION
**Pull request by yasinpalizban**
Supersedes #7247

**Description**
i add  cors  middleware  to  default filter CI4 and there  is a config file which can easily modify settings. 
all we have to do  go to filter.php  and un comment  cors  that's all needs.

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
Thank you for the update.

You must GPG-sign your commits.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing

We need unit testing.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#unit-testing

File: app/Config/Filters.php [hunk 1]
```diff
@@ -32,6 +34,7 @@ class Filters extends BaseConfig
             // 'honeypot',
             // 'csrf',
             // 'invalidchars',
+            // 'cors' check
```
Comment (kenjis) [inline app/Config/Filters.php:37]:
```suggestion
            // 'cors'
```

File: system/Filters/Cors.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            // get origin when it is  using Postman
+            $origin = $_SERVER['REMOTE_ADDR'];
+        }
+
+            // get  cors configuration
+        $corsConfig = config('Cors');
+
+        //  if origin is not in list , origin header no back in response object  and front end  framework get cors error
+        if (in_array($origin, $corsConfig->validDomains)) {
+            header('Access-Control-Allow-Origin: ' . $origin);
```
Comment (kenjis) [inline system/Filters/Cors.php:37]: Fix it so that header() is not used.

File: system/Filters/Cors.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        header("Access-Control-Allow-Credentials: {$corsConfig->credentials}");
+        // append set  max age
+        header("Access-Control-Max-Age: {$corsConfig->maxAge}");
+        // append content type
+        header('content-type: ' . implode("; ", $corsConfig->contentType));
+
+
+        // if request is option call back to browser 200 ok
+        if ($request->getMethod() == "options") {
+            header("HTTP/1.1 200 OK CORS");
```
Comment (kenjis) [inline system/Filters/Cors.php:55]: Fix it so that header() is not used.

File: system/Filters/Cors.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
```
Comment (kenjis) [inline system/Filters/Cors.php:56]: Please return a Response object.
See https://codeigniter4.github.io/CodeIgniter4/incoming/filters.html#before-filters

File: system/Filters/Cors.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    {
+
+
+            // get origin
+        if (array_key_exists('HTTP_ORIGIN', $_SERVER)) {
+            //get origin when it is using  front end framework angular or react
+            $origin = $_SERVER['HTTP_ORIGIN'];
+        } else if (array_key_exists('HTTP_REFERER', $_SERVER)) {
+            //  get origin when it is not using  front end  only browser
+            $origin = $_SERVER['HTTP_REFERER'];
```
Comment (kenjis) [inline system/Filters/Cors.php:26]: Can you explain why do you use `HTTP_REFERER` as origin?

File: system/Filters/Cors.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        } else {
+            // get origin when it is  using Postman
+            $origin = $_SERVER['REMOTE_ADDR'];
```
Comment (kenjis) [inline system/Filters/Cors.php:29]: Can you explain why do you use `REMOTE_ADDR` as origin?

**Comment by kenjis** [comment]
This library would be a good reference for configuration.
https://github.com/fruitcake/php-cors
I am not sure if we need all the features.

**Comment by cijagani** [comment]
Here is one more example
https://github.com/agungsugiarto/codeigniter4-cors

@agungsugiarto may help on this

**Comment by kenjis** [comment]
@cijagani The package are using another vendor's namespace `Fluent`.
https://packagist.org/packages/fluent/
So I don't recommend it.

**Comment by cijagani** [comment]
Agree. We should consider as reference or we should compare features.

**Pull request closed without merging**